data/tutorials/language/0lg_00_basic_datatypes.md

```diff
@@ -0,0 +1,844 @@
+  | Merge_head -> "MERGE_HEAD";;
+val commit_to_string' : commit -> string = <fun>
+```
+
+The `match … with …` construct needs to be passed an expression that is inspected. The `function …` is a special form of anonymous function taking a parameter and forwarding it to a `match … with …` construct as shown above.
+
+Warning: Wrapping product types with parenthesis turns them into a single parameter.
+```ocaml
+# type t1 = T1 of int * bool;;
+type t1 = T1 of int * bool
+
+# type t2 = T2 of (int * bool);;
+type t2 = T2 of (int * bool)
+
+# let p = (4, false);;
+val p : int * bool = (4, false)
+
+# T1 p;;
+Error: The constructor T1 expects 2 argument(s),
+       but is applied here to 1 argument(s)
+
+# T2 p;;
+- : t2 = T2 (4, false)
+```
+
+The constructor `T1` has two parameters of type `int` and `bool` whilst the constructor `T2` has a single parameter of type `int * bool`.
+
+#### Recursive Variants
+
+A variant definition referring to itself is recursive. A constructor may wrap data from the type being defined.
+
+This is the case for the following definition, which can be used to store JSON values.
+```ocaml
+# type json =
+  | Null
+  | Bool of bool
+  | Int of int
+  | Float of float
+  | String of string
+  | Array of json list
+  | Object of (string * json) list;;
+type json =
+    Null
+  | Bool of bool
+  | Int of int
+  | Float of float
+  | String of string
+  | Array of json list
+  | Object of (string * json) list
+```
+
+Both constructors `Array` and `Object` contain values of type `json`.
+
+Functions defined using pattern matching on recursive variants are often recursive too. This function checks if a name is present in a whole JSON tree:
+```ocaml
+# let rec has_field name = function
+  | Array u ->
+      List.fold_left (fun b obj -> b || has_field name obj) false u
+  | Object u ->
+      List.fold_left (fun b (key, obj) -> b || key = name || has_field name obj) false u
+  | _ -> false;;
+val has_field : string -> json -> bool = <fun>
+```
+
+Here, the last pattern uses the symbol `_`, which catches everything. It allows returning `false` on all data which is neither `Array` nor `Object`.
+
+### Polymorphic Data Types
+
+#### Revisiting Predefined Types
+
+The predefined type `option` is defined as a variant type with two constructors: `Some` and `None`. It can contain values of any type, such as `Some 42` or `Some "hola"`. In that sense, `option` is polymorphic. Here is how it is defined in the standard library:
+
+```ocaml
+# #show option;;
+type 'a option = None | Some of 'a
+```
+
+The predefined type `list` is polymorphic in the same sense. It is a variant with two constructors and can hold data of any type. Here is how it is defined in the standard library:
+```ocaml
+# #show list;;
+type 'a list = [] | (::) of 'a * 'a list
+```
+
+The only bit of magic here is turning constructors into symbols. This is left unexplained in this tutorial. The types `bool` and `unit` also are regular variants, with the same magic:
```

Comment:
```suggestion
The only bit of magic here is turning constructors into symbols, which we don't cover in this tutorial. The types `bool` and `unit` also are regular variants, with the same magic:
```

If not in this tutorial, where can we find it?

Reply:
Need to add that into the “operators” tutorial